<?xml version="1.0"?>
<robot name="tiguan">
  <link name="base_link">
    <visual name="base_visual">
      <geometry>
        <mesh filename="package://tiguan_description/mesh/body.dae" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
  </link>

  <link name="rear_axle"/>

  <joint name="rear_axle_joint" type="fixed">
    <parent link="base_link"/>
    <child link="rear_axle"/>
    <origin xyz="0 0 0.35" rpy="0 0 0"/>
  </joint>

  <link name="vehicle_center"/>
  <joint name="vehicle_center_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="vehicle_center"/>
    <origin xyz="1.3 0 0" rpy="0 0 0"/>
  </joint>

  <link name="front_left_steer_link"/>
  <joint name="front_left_steer_link_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="front_left_steer_link"/>
    <origin xyz="2.734 0.815 0" rpy="0 0 0"/>
  </joint>

  <link name="front_right_steer_link"/>
  <joint name="front_right_steer_link_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="front_right_steer_link"/>
    <origin xyz="2.734 -0.815 0" rpy="0 0 0"/>
  </joint>

  <link name="rear_left_wheel_axle" />
  <link name="rear_right_wheel_axle" />

  <joint name="rear_axle_to_rear_left_wheel" type="fixed">
    <parent link="rear_axle"/>
    <child link="rear_left_wheel_axle"/>
    <origin xyz="0 0.815 0" rpy="0 0 0"/>
  </joint>

  <joint name="rear_axle_to_rear_right_wheel" type="fixed">
    <parent link="rear_axle"/>
    <child link="rear_right_wheel_axle"/>
    <origin xyz="0 -0.815 0" rpy="0 0 0"/>
  </joint>
  
  <link name="imu_link"/>
  <joint name="imu_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="imu_link"/>
    <origin xyz="0 0 0" rpy="3.1416 0 1.5708"/>
  </joint>
  
  <link name="gnss"/>

  <joint name="gnss_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="gnss"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <link name="front_radar"/>

  <joint name="front_radar_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="front_radar"/>
    <origin xyz="0.0 0.0 0.45" rpy="0 0 0"/>
  </joint>

  <link name="ussfrontrightside"/>

  <joint name="ussfrontrightside_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="ussfrontrightside"/>
    <origin xyz="3.2 -0.9 0.2" rpy="0 0 -1.57"/>
  </joint>  

  <link name="ussfrontleftside"/>

  <joint name="ussfrontleftside_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="ussfrontleftside"/>
    <origin xyz="3.2 0.9 0.2" rpy="0 0 1.57"/>
  </joint>  

  <link name="ussrearrightside"/>

  <joint name="ussrearrightside_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="ussrearrightside"/>
    <origin xyz="-0.4 -0.9 0.3" rpy="0 0 -1.57"/>
  </joint>  

  <link name="ussrearleftside"/>

  <joint name="ussrearleftside_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="ussrearleftside"/>
    <origin xyz="-0.4 0.9 0.3" rpy="0 0 1.57"/>
  </joint>  

  <link name="ussfrontleftmiddle"/>

  <joint name="ussfrontleftmiddle_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="ussfrontleftmiddle"/>
    <origin xyz="3.6 0.3 0.05" rpy="0 0 0"/>
  </joint>  

  <link name="ussfrontrightmiddle"/>

  <joint name="ussfrontrightmiddle_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="ussfrontrightmiddle"/>
    <origin xyz="3.6 -0.3 0.05" rpy="0 0 0"/>
  </joint>  

  <link name="ussfrontleft"/>

  <joint name="ussfrontleft_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="ussfrontleft"/>
    <origin xyz="3.45 0.7 0.15" rpy="0 0 0.785398"/>
  </joint>  

  <link name="ussfrontright"/>

  <joint name="ussfrontright_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="ussfrontright"/>
    <origin xyz="3.45 -0.7 0.15" rpy="0 0 -0.785398"/>
  </joint>  

  <link name="ussrearleftmiddle"/>

  <joint name="ussrearleftmiddle_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="ussrearleftmiddle"/>
    <origin xyz="-1.0 0.3 0.15" rpy="0 0 -3.14"/>
  </joint>  

  <link name="ussrearrightmiddle"/>

  <joint name="ussrearrightmiddle_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="ussrearrightmiddle"/>
    <origin xyz="-1.0 -0.3 0.15" rpy="0 0 -3.14"/>
  </joint>  

  <link name="ussrearright"/>

  <joint name="ussrearright_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="ussrearright"/>
    <origin xyz="-0.95 -0.7 0.15" rpy="0 0 -2.35619"/>
  </joint>  

  <link name="ussrearleft"/>

  <joint name="ussrearleft_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="ussrearleft"/>
    <origin xyz="-0.95 0.7 0.15" rpy="0 0 2.35619"/>
  </joint>

  <link name="2d_lidar"/>

  <joint name="2d_lidar_joint" type="fixed">
    <parent link="rear_axle"/>
    <child link="2d_lidar"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>

  <link name="lidar_left">
  </link>
  <joint name="base_link_to_lidar_left" type="fixed">
    <parent link="base_link"/>
    <child link="lidar_left"/>
    <origin xyz="0.75 +0.475 1.91" rpy="0.4 0.01 3.141593"/>
  </joint>

  <link name="lidar_right">
  </link>
  <joint name="lidar_left_to_lidar_right" type="fixed">
    <parent link="lidar_left"/>
    <child link="lidar_right"/>
    <origin xyz="-0.032 0.967 -0.375" rpy="0.772 0.007 3.141"/>
  </joint>

  <link name="lidar_center_sensor"/>
  <joint name="base_link_to_lidar_center_sensor" type="fixed">
    <parent link="base_link"/>
    <child link="lidar_center_sensor"/>
    <origin xyz="1.6 0.0 1.91" rpy="-0.017 -0.007 -0.06"/>
  </joint>

  <link name="camera_front_left"/>
  <joint name="camera_front_wide_to_camera_front_left_joint" type="fixed">
    <parent link="camera_front_wide"/>
    <child link="camera_front_left"/>
    <origin xyz="-0.424 0.003 -0.006" rpy="0.245 -1.519 -0.295"/>
  </joint>

  <link name="camera_front_narrow"/>
  <joint name="camera_front_wide_to_camera_front_narrow_joint" type="fixed">
    <parent link="camera_front_wide"/>
    <child link="camera_front_narrow"/>
    <origin xyz="0.12 0.005 -0.022" rpy="-0.072 -0.012 0.03"/>
  </joint>

  <link name="camera_front_right"/>
  <joint name="camera_front_wide_to_camera_front_right_joint" type="fixed">
    <parent link="camera_front_wide"/>
    <child link="camera_front_right"/>
    <origin xyz="0.532 0.005 -0.036" rpy="2.336 1.53 2.353"/>
  </joint>

  <link name="camera_back_right"/>
  <joint name="camera_front_wide_to_camera_back_right_joint" type="fixed">
    <parent link="camera_front_wide"/>
    <child link="camera_back_right"/>
    <origin xyz="0.477 -0.145 -1.453" rpy="2.753 1.014 3.05"/>
  </joint>

  <link name="camera_back"/>
  <joint name="camera_front_wide_to_camera_back_joint" type="fixed">
    <parent link="camera_front_wide"/>
    <child link="camera_back"/>
    <origin xyz="-0.026 -0.06 -1.978" rpy="2.971 -0.037 3.125"/>
  </joint>

  <link name="camera_back_left"/>
  <joint name="camera_front_wide_to_camera_back_left_joint" type="fixed">
    <parent link="camera_front_wide"/>
    <child link="camera_back_left"/>
    <origin xyz="-0.489 -0.15 -1.419" rpy="2.797 -1.078 -3.096"/>
  </joint>


</robot>
